### PR TITLE
drivers: usb: usb_dc_it82xx2: Optimize the USB driver and address bugs related to the FIFO and extended endpoint

### PR DESCRIPTION
**Pull request by Chenhongren**
Optimize the USB driver and address bugs related to the FIFO and extended endpoint.

### CODE REVIEW

**Review by tmon-nordic — DISMISSED** [review]

**Review by Dino-Li — DISMISSED** [review]

File: drivers/usb/device/usb_dc_it82xx2.c [hunk 1]
```diff
@@ -347,158 +305,176 @@ static int it82xx2_usb_fifo_ctrl(uint8_t ep)
...
-			bool set_clr)
-{
-	struct usb_it82xx2_regs *const usb_regs =
-		(struct usb_it82xx2_regs *)it82xx2_get_usb_regs();
-	struct epn_ext_ctrl_regs *epn_ext_ctrl =
-		usb_regs->fifo_regs[EP_EXT_REGS_DX].ext_0_3.epn_ext_ctrl;
+	switch (ctrl) {
+	case EP_STALL_SEND:
+		if (ep_idx % 2) {
+			epn0n1_ext_ctrl[idx].fields.epn1_send_stall_bit = enable;
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_it82xx2.c:332]: Please refactor it to get `fields` based on ep_idx and maybe other arguments.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: drivers/usb/device/usb_dc_it82xx2.c [hunk 2]
```diff
@@ -303,35 +302,48 @@ static void it8xxx2_usb_dc_wuc_init(const struct device *dev)
 	IRQ_CONNECT(IT8XXX2_WU90_IRQ, 0, it82xx2_wu90_isr, 0, 0);
 }
 
-/* Function it82xx2_get_ep_fifo_ctrl_reg_idx(uint8_t ep_idx):
+/**
+ * @brief Control it82xx2 usb fifo registers
```
[the remark this answers is not in the document]
Comment (Chenhongren) [reply drivers/usb/device/usb_dc_it82xx2.c]: Removed. Thanks.

File: drivers/usb/device/usb_dc_it82xx2.c [hunk 3]
```diff
@@ -303,35 +302,48 @@ static void it8xxx2_usb_dc_wuc_init(const struct device *dev)
...
-	uint8_t reg_idx = (ep_idx < EP8) ?
-		((ep_fifo_res[ep_idx % FIFO_NUM] - 1) * 2) :
-		((ep_fifo_res[ep_idx % FIFO_NUM] - 1) * 2 + 1);
-
-	return reg_idx;
-}
+	if (ep_idx == EP0) {
+		LOG_ERR("Invalid endpoint 0x%x", ep);
+		ret = -EINVAL;
+	} else {
```
[the remark this answers is not in the document]
Comment (Chenhongren) [reply drivers/usb/device/usb_dc_it82xx2.c]: Thanks for your suggestion. It has been resolved.

File: drivers/usb/device/usb_dc_it82xx2.c [hunk 4]
```diff
@@ -44,7 +44,6 @@ LOG_MODULE_REGISTER(usb_dc_it82xx2, CONFIG_USB_DRIVER_LOG_LEVEL);
 #define RX_LINE_RESET			0x00
```
[the remark this answers is not in the document]
Comment (Chenhongren) [reply drivers/usb/device/usb_dc_it82xx2.c:44]: The commit message has been rewritten.

File: drivers/usb/device/usb_dc_it82xx2.c [hunk 5]
```diff
@@ -934,7 +934,7 @@ int usb_dc_reset(void)
 	ff_regs[EP0].ep_rx_fifo_ctrl = FIFO_FORCE_EMPTY;
 	ff_regs[EP0].ep_tx_fifo_ctrl = FIFO_FORCE_EMPTY;
 
-	for (uint8_t idx = EP1; idx < EP4; idx++) {
+	for (int idx = EP1; idx < EP4; idx++) {
```
[the remark this answers is not in the document]
Comment (Chenhongren) [reply drivers/usb/device/usb_dc_it82xx2.c:903]: The IT82xx2 chip is a 32-bit MCU, and using 'int' improves performance.

File: drivers/usb/device/usb_dc_it82xx2.c [hunk 6]
```diff
@@ -450,7 +450,11 @@ static int it82xx2_usb_extend_ep_ctrl(uint8_t ep_idx,
 	if ((ctrl >= EXT_EP_DIR_IN) && (ctrl < EXT_EP_READY)) {
 		/* From EXT_EP_DIR_IN to EXT_EP_DISABLE */
 		reg_idx = ep_fifo_res[ep_idx % FIFO_NUM];
-		mask = 1 << (ext_ep_bit_shift[ep_idx - 4] * 2 + 1);
+		if (ctrl == EXT_EP_DIR_IN || ctrl == EXT_EP_DIR_OUT) {
+			mask = 1 << (ext_ep_bit_shift[ep_idx - 4] * 2 + 1);
```
[the remark this answers is not in the document]
Comment (Chenhongren) [reply drivers/usb/device/usb_dc_it82xx2.c]: The issue has been fixed based on your suggestion. Thank you.

File: drivers/usb/device/usb_dc_it82xx2.c (hunk 1 above)
Comment (Chenhongren) [reply drivers/usb/device/usb_dc_it82xx2.c]: Completed. Thanks.

File: drivers/usb/device/usb_dc_it82xx2.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (jfischer-no) [reply drivers/usb/device/usb_dc_it82xx2.c:903]: > The IT82xx2 chip is a 32-bit MCU, and using 'int' improves performance.

Nonsense. Btw, where are EP0, EP1 and EP4 defined? How can it be an improvement to replace a 0 or 1 with something non-obvious like EP0/EP1?
NACK to all obvious number to obscure EP? macro replacements.

**Review by jfischer-no — CHANGES_REQUESTED** [review]
NACK to all obvious number to obscure EP? macro replacements.

**Comment by Chenhongren** [comment]
> NACK to all obvious number to obscure EP? macro replacements.

For the 'int' modification, we would like to align data type of idx variable with length of the CPU's GPR.
With the alignment, compiler should be able to compile the variable optimally.
EPx definitions are in "soc/riscv/riscv-ite/common/chip_chipregs.h."
This change is to enhance readability since the ep index is frequently used for registers access(FIFO and EP registers).
Should we revert these two modifications?

**Comment by jfischer-no** [comment]
> > NACK to all obvious number to obscure EP? macro replacements.
> 
> For the 'int' modification, we would like to align data type of idx variable with length of the CPU's GPR. With the alignment, compiler should be able to compile the variable optimally.

https://github.com/zephyrproject-rtos/zephyr/pull/65497/files#diff-ff0289674414ba9557a8342774a72345ba0086f05a16cb2f16329c6be87eb973R903, `for (uint8_t idx = 1u; idx < 4u; idx++) {` is fine and valid code, and you compiler will compile it to the same as `for (unsigned int idx = 1u; idx < 4u; idx++) {` or `for (int idx = 1; idx < 4; idx++) {`. 

However https://github.com/zephyrproject-rtos/zephyr/pull/65497/files#diff-ff0289674414ba9557a8342774a72345ba0086f05a16cb2f16329c6be87eb973R1235, where you changed `uint32_t idx` to `int idx` is just wrong because data_len is uint32_t.

"compiler should be able to compile the variable optimally" should be the last thing you need to worry about.

>EPx definitions are in "soc/riscv/riscv-ite/common/chip_chipregs.h." This change is to enhance readability since the ep index is frequently used for registers access(FIFO and EP registers). Should we revert these two modifications?

EPx macros do not enhance anything and are rather confusing. I really do not want to grep some other fancy HAL just to get type and obvious numbers.

**Comment by Chenhongren** [comment]
Thanks for your guidance.
This update reverts the changes about the ```int``` and ```EPx``` modifications. Additionally, the previous definition of ```MAX_NUM_ENDPOINTS``` has been removed.
Please help review it again.

**Review by jfischer-no — APPROVED** [review]

**Review by tmon-nordic — APPROVED** [review]

**Pull request merged**